### PR TITLE
Optimize fixture generation

### PR DESCRIPTION
**Pull request by esoergel**
https://dimagi-dev.atlassian.net/browse/HI-663

There's more work to be done, but I figure this is a good start and it'd be nice to get the reindex out as early as possible.  It reindexes a view in the fixtures db, which looks to contain about 7M docs on prod.  My intuition is that that's small enough that it shouldn't be a huge concern, though I'll of course shepherd the roll-out.
The main thing this does is sorts fixture data items in the DB.  I'm very interested in suggestions for next steps.  Things I'd like to explore are:
* Keeping this an iterator as long as possible without evaluating and hopefully never holding entirely in memory
* Chunking the couch view call
* Possibly not wrapping the individual data items

Though I would like to push secondary changes into separate PRs to get this in quickly.

### CODE REVIEW

File: corehq/apps/fixtures/dbaccessors.py [hunk 1]
```diff
@@ -29,15 +29,15 @@ def get_fixture_data_types_in_domain(domain):
...
-@quickcache(['domain', 'data_type_ids'], timeout=60 * 60, memoize_timeout=60, skip_arg='bypass_cache')
-def get_fixture_items_for_data_types(domain, data_type_ids, bypass_cache=False):
+@quickcache(['domain', 'data_type_id'], timeout=60 * 60, memoize_timeout=60, skip_arg='bypass_cache')
+def get_fixture_items_for_data_type(domain, data_type_id, bypass_cache=False):
     from corehq.apps.fixtures.models import FixtureDataItem
     return list(FixtureDataItem.view(
         'fixtures/data_items_by_domain_type',
-        keys=[[domain, id] for id in data_type_ids],
+        startkey=[domain, data_type_id],
+        endkey=[domain, data_type_id, {}],
```
Comment (esoergel) [inline corehq/apps/fixtures/dbaccessors.py]: Note that the old key format is not forward compatible.  Calls to this between migration and code deploy will fail.  Might end up making an accommodation for that if necessary

**Review by biyeun — APPROVED** [review]

**Review by millerdev — APPROVED** [review]

File: corehq/apps/fixtures/fixturegenerators.py [hunk 2]
```diff
@@ -116,46 +116,38 @@ def get_global_items(self, global_types, restore_state):
...
-            return item.sort_key
-
-        items_by_type = dict(items_by_type)
-        for data_type in data_types.values():
-            if data_type not in items_by_type:
-                items_by_type[data_type] = []
+    def _get_fixtures(self, data_types, get_items_by_type, user_id):
         fixtures = []
-        for data_type, items in sorted(list(items_by_type.items()), key=tag):
+        for data_type in sorted(data_types.values(), key=lambda data_type: data_type.tag):
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/fixtures/fixturegenerators.py]: oh neat, I didn't know about that, thanks!

File: corehq/apps/fixtures/fixturegenerators.py [hunk 3]
```diff
@@ -116,46 +116,38 @@ def get_global_items(self, global_types, restore_state):
...
-            items_by_type[data_type].append(item)
-        return self._get_fixtures(user_types, items_by_type, restore_user.user_id)
+            data_type = user_types.get(item.data_type_id)
+            if data_type:
+                self._set_cached_type(item, data_type)
+                items_by_type[data_type].append(item)
+
+        def get_items_by_type(data_type):
+            return sorted(items_by_type.get(data_type, []),
+                          key=lambda item: item.sort_key)
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/fixtures/fixturegenerators.py]: That's a good point - I don't believe it's possible, as this is called once per value in `user_types`, which is a `data_type_id: data_type` dict.  It's definitely not obvious from the structure of the code :/

Still, I'd like to encourage this code to feel like a once-through stream (in line with the version created in `_get_global_items`) and switch as much of this as possible to a lazily evaluated iterable that isn't held in memory.  The user fixtures aren't much of a performance concern (at least not yet), so I just did what was simplest to keep them parallel to the global fixtures

**Comment by esoergel** [comment]
I pulled out the first part of this into https://github.com/dimagi/commcare-hq/pull/24493 to allow for an intermediate state where the code is compatible with both versions of the view.

**Comment by esoergel** [comment]
Forgot to mention this explicitly, but another notable change in here is that the restore code now skips the quickcache on `get_fixture_items_for_data_type`.  That was trying to store all fixture items in a single quickcache entry, which could very plausibly have been a point of failure.  Since this PR switches it to pull in chunks, caching doesn't really work, and the fixtures restore code has that custom caching mechanism anyways.

File: corehq/apps/fixtures/tests/test_fixture_data.py [hunk 4]
```diff
@@ -330,3 +330,71 @@ def make_data_item(self, data_type, cost):
...
+    def _make_data_item(cls, sort_key, name, seat, sigil):
+        def field_list(value):
+            return FieldList(field_list=[FixtureItemField(field_value=value, properties={})])
+
+        data_item = FixtureDataItem(
+            domain=cls.domain,
+            data_type_id=cls.data_type._id,
+            fields={
+                "name": field_list(name),
+                "set": field_list(seat),
```
Comment (proteusvacuum) [inline corehq/apps/fixtures/tests/test_fixture_data.py]: nit: did you mean `seat` here?

**Review by proteusvacuum — COMMENTED** [review]
nice!

File: corehq/apps/fixtures/tests/test_fixture_data.py (hunk 4 above)
Comment (esoergel) [reply corehq/apps/fixtures/tests/test_fixture_data.py]: https://github.com/dimagi/commcare-hq/pull/24648

**Pull request merged**